nipype/algorithms/confounds.py

```diff
@@ -1201,20 +1281,56 @@ def compute_noise_components(imgseries, mask_images, num_components,
         # "The covariance matrix C = MMT was constructed and decomposed into its
         # principal components using a singular value decomposition."
         try:
-            u, _, _ = fallback_svd(M, full_matrices=False)
+            u, s, _ = fallback_svd(M, full_matrices=False)
         except np.linalg.LinAlgError:
-            if self.inputs.failure_mode == 'error':
+            if failure_mode == 'error':
                 raise
-            u = np.ones((M.shape[0], num_components), dtype=np.float32) * np.nan
+            if components_criterion >= 1:
+                u = np.empty((M.shape[0], components_criterion),
+                             dtype=np.float32) * np.nan
+            else:
+                continue
+
+        variance_explained = (s ** 2) / np.sum(s ** 2)
+        cumulative_variance_explained = np.cumsum(variance_explained)
+
+        num_components = int(components_criterion)
+        if 0 < components_criterion < 1:
+            num_components = np.searchsorted(cumulative_variance_explained,
+                                             components_criterion) + 1
+        elif components_criterion == -1:
+            num_components = len(s)
+
+        num_components = int(num_components)
+        if num_components == 0:
+            break
         if components is None:
             components = u[:, :num_components]
+            metadata = OrderedDict()
+            metadata['mask'] = [name] * len(s)
+            metadata['singular_value'] = s
+            metadata['variance_explained'] = variance_explained
+            metadata['cumulative_variance_explained'] = (
+                cumulative_variance_explained)
+            metadata['retained'] = [i < num_components for i in range(len(s))]
         else:
             components = np.hstack((components, u[:, :num_components]))
-    if components is None and num_components > 0:
-        if self.inputs.failure_mode == 'error':
+            metadata['mask'] = metadata['mask'] + [name] * len(s)
+            metadata['singular_value'] = (
+                np.hstack((metadata['singular_value'], s)))
+            metadata['variance_explained'] = (
+                np.hstack((metadata['variance_explained'],
+                           variance_explained)))
+            metadata['cumulative_variance_explained'] = (
+                np.hstack((metadata['cumulative_variance_explained'],
+                           cumulative_variance_explained)))
+            metadata['retained'] = (
+                metadata['retained'] + [i < num_components for i in range(len(s))])
+    if components is None:
+        if failure_mode == 'error':
             raise ValueError('No components found')
-        components = np.ones((M.shape[0], num_components), dtype=np.float32) * np.nan
-    return components, basis
+        components = np.full((M.shape[0], num_components), np.NaN)
```

Comment:
```suggestion
        components = np.full((M.shape[0], num_components), np.nan)
```

See [np.NaN](https://docs.scipy.org/doc/numpy/reference/constants.html?highlight=nan#numpy.NaN).